File: search_and_sort/shell_sort.lua
local function shell_sort(arr, len)
	local gap = len / 2
	while gap > 0 do
		for i = gap + 1, len do
			local temp = arr[i]
			local j = i
			while j > gap and arr[j - gap] > temp do
				arr[j] = arr[j - gap]
				j = j - gap
			end
			arr[j] = temp
		end
		gap = math.floor(gap / 2)
	end
end

local function main()
	local arr = {}
	for i = 1, N do
		arr[i] = math.random(0, MAX_ELEMENT)
	end
	shell_sort(arr, N)
	print(table.concat(arr, ", "))
end

N = 100
MAX_ELEMENT = 10000

main()

File: search_and_sort/shell_sort.py

from random import randint


N: int = 100
MAX_ELEMENT: int = 10000


def shell_sort(arr):
    gap = len(arr) // 2
    while gap > 0:
        for i in range(gap, len(arr)):
            j = i
            while j >= gap and arr[j - gap] > arr[j]:
                arr[j - gap], arr[j] = arr[j], arr[j - gap]
                j -= gap
        gap //= 2
    return arr


if __name__ == '__main__':
    arr = [randint(0, MAX_ELEMENT) for _ in range(N)]
    print(shell_sort(arr))


